README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# klexdatr

<!-- badges: start -->

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![R build
status](https://github.com/poissonconsulting/klexdatr/workflows/R-CMD-check/badge.svg)](https://github.com/poissonconsulting/klexdatr/actions)
[![License:
CCBY4](https://img.shields.io/badge/License-CCBY4-green.svg)](https://creativecommons.org/licenses/by/4.0/)
[![CRAN
status](https://www.r-pkg.org/badges/version/klexdatr)](https://CRAN.R-project.org/package=klexdatr)
<!-- badges: end -->

`klexdatr` is an R package of six relational tibbles from the Kootenay
Lake Large Trout Exploitation study.

``` r
ls("package:klexdatr") # the six tibbles
#> [1] "capture"    "deployment" "detection"  "recapture"  "section"   
#> [6] "station"
```

``` r
library(tibble) # for prettier printing
capture
#> # A tibble: 245 x 9
#>    Capture DateTimeCapture     SectionCapture Species    Length Weight Reward1
#>    <fct>   <dttm>              <fct>          <fct>       <int>  <dbl>   <int>
#>  1 F075    2008-05-24 10:40:00 S25            Bull Trout    545   1.75       0
#>  2 F076    2008-05-24 11:10:00 S25            Bull Trout    654   3.4        0
#>  3 F084    2008-06-05 08:01:00 S16            Bull Trout    553  NA          0
#>  4 F100    2009-05-02 13:31:00 S21            Bull Trout    643   2.75     100
#>  5 F104    2009-05-03 11:17:00 S22            Bull Trout    567   2        100
#>  6 F129    2009-05-07 09:11:00 S21            Bull Trout    703   3.9      100
#>  7 F133    2009-05-07 13:58:00 S21            Bull Trout    712   3.65     100
#>  8 F137    2009-05-08 11:16:00 S21            Bull Trout    572   1.9      100
#>  9 F141    2009-05-20 15:13:00 S16            Bull Trout    564   2.2      100
#> 10 F143    2009-05-21 08:00:00 S16            Bull Trout    595   2.4      100
#> # … with 235 more rows, and 2 more variables: Reward2 <int>,
#> #   DateTimeTagExpire <dttm>
```

``` r
recapture
#> # A tibble: 54 x 8
#>    DateTimeRecapture   Capture SectionRecapture TBarTag1 TBarTag2 TagsRemoved
#>    <dttm>              <fct>   <fct>            <lgl>    <lgl>    <lgl>      
#>  1 2009-04-05 12:00:00 F006    S25              TRUE     TRUE     TRUE       
#>  2 2009-05-23 12:00:00 F131    S16              TRUE     TRUE     TRUE       
#>  3 2009-09-10 12:00:00 F097    S14              TRUE     TRUE     TRUE       
#>  4 2009-10-06 12:00:00 F169    S08              TRUE     TRUE     TRUE       
#>  5 2009-10-10 12:00:00 F076    S30              TRUE     TRUE     TRUE       
#>  6 2009-10-12 12:00:00 F099    S10              TRUE     TRUE     TRUE       
#>  7 2009-10-14 12:00:00 F112    S16              TRUE     TRUE     TRUE       
#>  8 2009-10-16 12:00:00 F111    S27              TRUE     TRUE     TRUE       
#>  9 2010-01-23 12:00:00 F197    S27              TRUE     TRUE     TRUE       
#> 10 2010-01-28 12:00:00 F075    S23              TRUE     TRUE     TRUE       
#> # … with 44 more rows, and 2 more variables: Released <lgl>, Public <lgl>
```

## Installation

To install the latest release from [CRAN](https://cran.r-project.org)

``` r
install.packages("klexdatr")
```

To install the developmental version from
[GitHub](https://github.com/poissonconsulting/klexdatr)

``` r
# install.packages("remotes")
remotes::install_github("poissonconsulting/klexdatr")
```

## Information

The study which ran from 2008 to 2014 caught, tagged and released large
Rainbow Trout and Bull Trout in Kootenay Lake by boat angling.

The fish were tagged with internal acoustic tags and/or high reward
external tags and subsequently detected by an acoustic receiver array as
well as reported by anglers.

The data are analysed by [Thorley and Andrusak
(2017)](https://peerj.com/articles/2874/) to estimate the natural and
fishing mortality of both species.

There is also a [shiny
app](https://www.poissonconsulting.ca/post/2018/klexdatr/) to view the
fish movement data.

## Contribution

Please report any
[issues](https://github.com/poissonconsulting/klexdatr/issues).

[Pull requests](https://github.com/poissonconsulting/klexdatr/pulls) are
always welcome.

## Code of Conduct

Please note that the klexdatr project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## Acknowledgements

![](man/figures/logos.png)

The project was primarily funded by the Habitat Conservation Trust
Foundation.

The Habitat Conservation Trust Foundation was created by an act of the
legislature to preserve, restore and enhance key areas of habitat for
fish and wildlife throughout British Columbia. Anglers, hunters,
trappers and guides contribute to the projects of the Foundation through
licence surcharges. Tax deductible donations to assist in the work of
the Foundation are also welcomed.

The project was partially funded by the Fish and Wildlife Compensation
Program on behalf of its program partners BC Hydro, the Province of
B.C., Fisheries and Oceans Canada, First Nations and the public who work
together to conserve and enhance fish and wildlife impacted by the
construction of BC Hydro dams.

Annual operation and maintenance for VR2W arrays used in this study were
completed by Ministry of Forests, Lands and Natural Resource Operations
(MFLNRO) and funded by the Fish and Wildlife Compensation Program (FWCP)
in conjunction with the Bonneville Power Administration (BPA) through
the Northwest Power and Conservation Council’s Fish and Wildlife
Program, in co-operation with the Idaho Department of Fish and Game
(IDFG), and the Kootenai Tribe of Idaho (KTOI). The Freshwater Fish
Society of British Columbia (FFSBC) provided tag rewards


NEWS.md
<!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->

# klexdatr 0.1.2.9000

- Same as previous version.


# klexdatr 0.1.2

- Converted `station` and `section` to `sf` objects (not just with `sfc` vectors).
- Imported `sf` package so no problem printing `sfc` vectors.
- No longer test with R 3.3.

# klexdatr 0.1.1

- Internal changes only.

# klexdatr 0.1.0

- Converted data sets to tibbles with sfc vectors.

# klexdatr 0.0.8

- Updated Documentation
      
# klexdatr 0.0.7

- Used in Thorley and Andrusak (2017)

# klexdatr 0.0.6

- Code rivers and West Arm as Lotic.
- Code West Arm and Kootenay River as not Bounded.

# klexdatr 0.0.5

- Keep all captures with acoustic tag and/or $100 reward

# klexdatr 0.0.4

- Exported lexr plotting functions in README

# klexdatr 0.0.3

- Removed depth data as no resources to validate

# klexdatr 0.0.2

- Added reward fish without acoustic tags

# klexdatr 0.0.1

- Initial release


tests/testthat.R
library(testthat)
library(qlexdatr)

test_check("qlexdatr")


tests/testthat/test-data.R
test_that("data", {
  expect_null(chk::check_data(
    section,
    list(
      Section = factor(1),
      Habitat = factor(c("Lentic", "Lentic", "Lotic"),
        levels = c("Lentic", "Lotic")
      ),
      Bounded = c(TRUE, TRUE, FALSE)
    ),
    key = "Section"
  ))

  expect_null(chk::chk_is(section, "sf"))

  expect_null(chk::check_data(
    station,
    list(
      Station = factor(1),
      Section = factor(1)
    ),
    key = "Station"
  ))

  expect_null(chk::chk_is(station, "sf"))

  expect_null(chk::chk_join(tibble::as_tibble(station),
                            tibble::as_tibble(section), "Section"))

  expect_null(chk::check_data(
    deployment,
    list(
      Station = factor(1),
      Receiver = factor(1),
      DateTimeReceiverIn = Sys.time(),
      DateTimeReceiverOut = Sys.time()
    ),
    key = c("Station", "Receiver", "DateTimeReceiverIn")
  ))

  expect_null(chk::chk_join(deployment, tibble::as_tibble(station), "Station"))

  expect_null(chk::check_data(
    capture,
    list(
      Capture = factor(1),
      DateTimeCapture = Sys.time(),
      SectionCapture = factor(1),
      Species = factor(""),
      Length = c(200L, 1000L),
      Weight = c(0.5, 10, NA),
      Reward1 = c(0L, 10L, 100L),
      Reward2 = c(0L, 10L, 100L, NA),
      DateTimeTagExpire = Sys.time()
    ),
    key = c("Capture")
  ))

  expect_null(chk::chk_join(capture, tibble::as_tibble(section), c("SectionCapture" = "Section")))

  expect_null(chk::check_data(
    recapture,
    list(
      DateTimeRecapture = Sys.time(),
      Capture = factor(1),
      SectionRecapture = factor(c(1, NA)),
      TBarTag1 = TRUE,
      TBarTag2 = TRUE,
      TagsRemoved = TRUE,
      Released = TRUE,
      Public = TRUE
    ),
    key = c("DateTimeRecapture", "Capture")
  ))

  expect_null(chk::chk_join(recapture, capture, "Capture"))
  #  missing values
  #  expect_null(chk::chk_join(recapture, section@data,
  #                                   c("SectionRecapture" = "Section")))

  expect_null(chk::check_data(
    detection,
    list(
      DateTimeDetection = Sys.time(),
      Capture = factor(1),
      Receiver = factor(1),
      Detections = c(1L, 300L)
    ),
    key = c("DateTimeDetection", "Capture", "Receiver")
  ))

  expect_null(chk::chk_join(detection, capture, "Capture"))
})


R/data.R
#' Section Data
#'
#' Section Spatial Polygon Data
#'
#' Polygons of sections of the waterbodies.
#'
#' @format A SpatialPolygonsDataFrame with the data frame:
#' \describe{
#'   \item{Section}{The unique section code (fctr).}
#'   \item{Habitat}{The habitat type 'Lentic' or 'Lotic' (fctr).}
#'   \item{Bounded}{The polygon represents the full area (lgl).}
#'   \item{geometry}{The section polygon (MULTIPOLYGON (m)).}
#' }
"section"

#' Station Data
#'
#' A tbl data frame of detection stations.
#'
#' @format A tbl data frame:
#' \describe{
#'   \item{Station}{The unique station name (fctr).}
#'   \item{Section}{The section code (fctr).}
#'   \item{geometry}{The station point (POINT (m)).}
#' }
"station"

#' Receiver Deployment Data
#'
#' A data frame of receiver deployments by station and date times.
#'
#' @format A tbl data frame:
#' \describe{
#'   \item{Station}{The station name (fctr).}
#'   \item{Receiver}{The receiver code (fctr).}
#'   \item{DateTimeReceiverIn}{The receiver deployment date and time (time).}
#'   \item{DateTimeReceiverOut}{The receiver retrieval date and time (time).}
#' }
"deployment"

#' Fish Capture Data
#'
#'
#' @format A tbl data frame:
#' \describe{
#'   \item{Capture}{The unique fish code (fctr).}
#'   \item{DateTimeCapture}{The date and time of capture (time).}
#'   \item{SectionCapture}{The section code (fint).}
#'   \item{Species}{The fish species 'Bull Trout', 'Lake Trout' or 'Rainbow Trout' (fctr).}
#'   \item{Length}{The fork length in mm (int).}
#'   \item{Weight}{The wet mass in kg (dbl).}
#'   \item{Reward1}{The reward value of the first T-Bar tag in Canadian dollars (int).}
#'   \item{Reward2}{The reward value of the second T-Bar tag if present in Canadian dollars (int).}
#'   \item{DateTimeTagExpire}{The acoustic tag expiration date and time (time).}
#' }
"capture"

#' Fish Recapture Data
#'
#' A tbl data frame of fish recaptures. As the time of recapture
#' was not reported it is assumed to be 12:00:00.
#'
#' @format A tbl data frame:
#' \describe{
#'   \item{DateTimeRecapture}{The reported date of recapture (time).}
#'   \item{Capture}{The fish code (fctr).}
#'   \item{SectionRecapture}{The section code (fctr).}
#'   \item{TBarTag1}{The first T-Bar Tag was reported (lgl).}
#'   \item{TBarTag2}{A second T-Bar Tag was reported (lgl).}
#'   \item{TagsRemoved}{The T-Bar tags were removed from the fish (lgl).}
#'   \item{Released}{The angler reportedly released the fish (lgl).}
#'   \item{Public}{The angler was a member of the public as opposed the study team (lgl).}
#' }
"recapture"

#' Acoustic Detection Data
#'
#' Hourly acoustic detection data by fish (capture) and receiver.
#'
#' @format A tbl data frame:
#' \describe{
#'   \item{DateTimeDetection}{The detection date and hour (time).}
#'   \item{Capture}{The fish code (fctr).}
#'   \item{Receiver}{The receiver code (fctr).}
#'   \item{Detections}{The number of detections in the hour (int).}
#' }
"detection"


scripts/dm.R
remotes::install_github("krlmlr/dm#335")

library(tidyverse)
library(dm)

pkgload::load_all()

dm <-
  dm(
    station,
    section,
    capture,
    deployment,
    detection,
    recapture
  )

dm %>%
  dm_add_pk(capture, Capture) %>%
  dm_add_pk(deployment, c(Station, Receiver, DateTimeReceiverIn)) %>%
  dm_add_pk(detection, c(DateTimeDetection, Capture, Receiver)) %>%
  dm_add_pk(recapture, c(DateTimeRecapture, Capture)) %>%
  dm_add_pk(section, Section) %>%
  dm_add_pk(station, Station) %>%
  dm_draw(rankdir = "TB", view_type = "all")


dm_keyed <-
  dm %>%
  dm_add_pk(capture, Capture) %>%
  dm_add_pk(deployment, c(Station, Receiver, DateTimeReceiverIn)) %>%
  dm_add_pk(detection, c(DateTimeDetection, Capture, Receiver)) %>%
  dm_add_pk(recapture, c(DateTimeRecapture, Capture)) %>%
  dm_add_pk(section, Section) %>%
  dm_add_pk(station, Station) %>%
  dm_add_fk(detection, Capture, capture) %>%
  dm_add_fk(recapture, Capture, capture) %>%
  dm_add_fk(recapture, SectionRecapture, section) %>%
  dm_add_fk(capture, SectionCapture, section) %>%
  dm_add_fk(deployment, Station, station) %>%
  dm_add_fk(station, Section, section)

dm_keyed %>%
  dm_draw()

dm_keyed %>%
  dm_flatten_to_tbl(deployment)


data-raw/data-raw.R
library(sp)
library(spdep)
library(spa)
library(rgdal)
library(rgeos)
library(raster)
library(maptools)
library(readr)
library(plyr)
library(dplyr)
library(magrittr)
library(lubridate)
library(assertr)
library(devtools)
library(checkr)
library(lexr)

rm(list = ls())

dir <- path.expand("~/Dropbox/Data")
dir %<>% file.path(basename(getwd())) %>%
  file.path("20160201")

epsg_data <- 26911
epsg_analysis <- 26911 # UTM Zone 11U
tz_data <- "PST8PDT"
tz_analysis <- "Etc/GMT+8"
species <- list("Bull Trout" = "BT", "Rainbow Trout" = "RB")

firstDateTime <- as.POSIXct("2008-04-01 00:00:00", tz = tz_analysis)
lastDateTime <- as.POSIXct("2013-12-31 23:59:59", tz = tz_analysis)

section <- readOGR(dsn = file.path(dir, "Shape"), layer = "array2015section")
section %<>% spTransform(CRS(paste0("+init=epsg:", epsg_analysis)))

section@data %<>% select(SectionNumber = SECTNBR) %>% check_key("SectionNumber")

section@data %<>% mutate(Section = paste0("S", sprintf("%02d", SectionNumber)))

# combine sections
section@data$Section[section@data$SectionNumber %in% 33:38] <- "S33"

section@data <- as.data.frame(bind_cols(
  section@data, select(as.data.frame(gCentroid(section, byid = TRUE)),
                       EastingSection = x, NorthingSection = y)))

section <- section[order(section@data$NorthingSection * -1,
                         section@data$EastingSection),]

section@data$Section %<>% factor(levels = unique(.))

lexr:::plot_lex_section(section)

deployment <- read_csv(file.path(dir, "qryKLESReceiversAnalysis22Dec2015.txt"))

deployment %<>% mutate(Station = SiteName2015,
                       DateTimeReceiverIn = ISOdate(InYear, InMonth, InDay, tz = tz_data),
                       DateTimeReceiverOut = ISOdate(OutYear, OutMonth, OutDay, , tz = tz_data)) %>%
  mutate(DateTimeReceiverIn = with_tz(DateTimeReceiverIn, tz_analysis),
         DateTimeReceiverOut = with_tz(DateTimeReceiverOut, tz_analysis))

deployment %<>% filter((DateTimeReceiverIn >= firstDateTime & DateTimeReceiverIn <= lastDateTime) | (DateTimeReceiverOut >= firstDateTime & DateTimeReceiverOut <= lastDateTime))

deployment$Reliable <- TRUE
message("no data several deployments 2008")
deployment$Reliable[deployment$Station %in% c(
  "Fry Creek", "South of Kaslo - West", "South of Kaslo - East")
  & deployment$InYear == 2008] <- FALSE

message("several apparently failed deployments")
deployment$Reliable[deployment$Station %in% c(
  "Woodbury Point")
  & deployment$InYear == 2009] <- FALSE

deployment$Reliable[deployment$Station %in% c(
  "Redman Point W")
  & deployment$InYear == 2010] <- FALSE

deployment$Reliable[deployment$Station == "South of Kaslo - West" & deployment$InYear == 2008 & deployment$InMonth == 5] <- FALSE

deployment %<>% filter(Reliable)

station <- select(deployment, Station, EastingStation = Xn83z11u, NorthingStation = Yn83z11u)

station %<>% group_by(Station) %>% summarise(EastingStation = mean(EastingStation), NorthingStation = mean(NorthingStation)) %>% ungroup()

station %<>% check_key("Station") %>% as.data.frame()

station <- SpatialPointsDataFrame(select(station, EastingStation, NorthingStation),
                                  station,
                                  proj4string = CRS(paste0("+init=epsg:", epsg_data)))

station %<>% spTransform(CRS(paste0("+init=epsg:", epsg_analysis)))
station@data[,c("EastingStation", "NorthingStation")] <- coordinates(station)

station <- bind_cols(station@data, select(sp::over(station, section), Section)) %>%
  filter(!is.na(Section))

# filter stations
station %<>% filter(!Section %in% c("S01", "S04", "S05", "S06", "S19", "S33"))

station %<>% arrange(Section, NorthingStation, EastingStation)
station$Station %<>% factor(., levels = .)
station %<>%  select(Station, Section, EastingStation, NorthingStation)

lexr:::plot_lex_station(station, section)
lexr:::check_lex_station(station)
use_data(station, overwrite = TRUE)

deployment %<>% filter(Station %in% station$Station)

deployment$Station %<>% factor(., levels = levels(station$Station))
deployment %<>% select(Station, Receiver = RecNbr, DateTimeReceiverIn, DateTimeReceiverOut) %>% as.tbl() %>% verify(DateTimeReceiverIn < DateTimeReceiverOut)

# filter overlapping deployments
deployment %<>% filter(!Receiver %in% c(0,2:5)) %>% filter(!Receiver %in% 3228:3231)
deployment$Receiver %<>% factor()

lexr:::plot_lex_deployment(deployment)
lexr:::check_lex_deployment(deployment)
use_data(deployment, overwrite = TRUE)

acoustic_tag <- read_csv(file.path(dir, "qryKLESAcousticTag22Dec2015.txt"))
acoustic_tag %<>% filter(!is.na(TagLife))
acoustic_tag %<>% select(AcousticTag = TagIDNbr, TagLife) %>%
  mutate(TagLife = as.integer(TagLife))

capture <- read_csv(file.path(dir, "qryKLESCaptureAnalysis22Dec2015.txt"))

capture %<>% mutate(
  DateTimeCapture = ISOdate(CapYear, CapMonth, CapDay, CapHour, CapMinute, tz = tz_data)) %>%
  mutate(DateTimeCapture = with_tz(DateTimeCapture, tz_analysis))

capture %<>% filter(DateTimeCapture >= firstDateTime & DateTimeCapture <= lastDateTime)

capture %<>% rename(AcousticTag = TagIDNbr) %>%
  left_join(acoustic_tag, by = "AcousticTag")

adjust_taglife <- function(x) {
  if (is.na(x$AcousticTag[1])) {
    x$TagLife <- 0
    return(x)
  }
  if (nrow(x) == 1)
    return(x)
  x %<>% arrange(DateTimeCapture)
  x$TagLife %<>% subtract(as.Date(x$DateTimeCapture) - as.Date(x$DateTimeCapture[1])) %>%
    as.integer()
  x
}

capture %<>% ddply(.(AcousticTag), adjust_taglife)

capture %<>% mutate(DateTimeTagExpire = DateTimeCapture + days(TagLife))

capture %<>% mutate(CaptureX = Xn83z11u,
                    CaptureY = Yn83z11u,
                    Species = factor(Species),
                    Length = as.integer(Length))

levels(capture$Species) <- species

capture %<>% filter(MortalityYN != "Yes")
capture %<>% filter(Length > 0)

is.na(capture$Weight[capture$WeigthType != "Measured"]) <- TRUE

capture %<>% select(Capture = CaptureID, DateTimeCapture, Species, AcousticTag,
                    Length, Weight,
                    TBarTag1, TBarTag1Reward, TBarTag2, TBarTag2Reward,
                    DateTimeTagExpire, CaptureX, CaptureY)

capture %<>% filter(!is.na(TBarTag1) | !is.na(TBarTag2))

capture %<>% mutate(Switch = !is.na(TBarTag2) & TBarTag1Reward < TBarTag2Reward)

tbartag <- capture$TBarTag2[capture$Switch]
capture$TBarTag2[capture$Switch] <- capture$TBarTag1[capture$Switch]
capture$TBarTag1[capture$Switch] <- tbartag

reward <- capture$TBarTag2Reward[capture$Switch]
capture$TBarTag2Reward[capture$Switch] <- capture$TBarTag1Reward[capture$Switch]
capture$TBarTag1Reward[capture$Switch] <- reward

capture %<>% select(-Switch) %>% verify(!is.na(TBarTag1)) %>% verify(is.na(TBarTag2Reward) | TBarTag1Reward >= TBarTag2Reward) %>% as.data.frame()

# only keep acoustic tag or fish with reward tags
capture %<>% filter(DateTimeTagExpire > DateTimeCapture | TBarTag1Reward >= 100)

capture <- SpatialPointsDataFrame(select(capture, CaptureX, CaptureY),
                                  select(capture, -CaptureX, -CaptureY),
                                  proj4string = CRS(paste0("+init=epsg:", epsg_data)))

capture %<>% spTransform(CRS(paste0("+init=epsg:", epsg_analysis)))
capture@data[,c("CaptureX", "CaptureY")] <- coordinates(capture)

capture <- bind_cols(capture@data, select(sp::over(capture, section), Section)) %>%
  filter(!is.na(Section))

recapture <- read_csv(file.path(dir, "qryKLESRecaptureAnalysis26Jan2016.txt"))

recapture %<>% mutate(
  DateTimeRecapture = ISOdate(ReCapYear, ReCapMonth, ReCapDay, tz = tz_analysis))

recapture %<>% filter(DateTimeRecapture >= firstDateTime & DateTimeRecapture <= lastDateTime)

recapture %<>% select(DateTimeRecapture,
                      Released, TagsRemoved,
                      TBarTag1Recap, TBarTag2Recap, Public = RecapType,
                      RecaptureX = Xn83z11u, RecaptureY = Yn83z11u,
                      Capture = CaptureID, RecaptureDBID = RecaptureID) %>%
  as.data.frame()

recapture$RecaptureX[is.na(recapture$RecaptureX)] <- 0
recapture$RecaptureY[is.na(recapture$RecaptureY)] <- 0

recapture <- SpatialPointsDataFrame(select(recapture, RecaptureX, RecaptureY),
                                    recapture,
                                    proj4string = CRS(paste0("+init=epsg:", epsg_data)))

recapture %<>% spTransform(CRS(paste0("+init=epsg:", epsg_analysis)))
recapture@data[,c("RecaptureX", "RecaptureY")] <- coordinates(recapture)

recapture <- bind_cols(recapture@data, select(sp::over(recapture, section), Section))

recapture$Released %<>% factor()
levels(recapture$Released) <- list("FALSE" = "No", "TRUE" = "Yes")
recapture$Released %<>% as.logical()

recapture$TagsRemoved %<>% factor()
levels(recapture$TagsRemoved) <- list("FALSE" = "No", "TRUE" = "Yes")
recapture$TagsRemoved %<>% as.logical()

recapture$Public %<>% factor()
levels(recapture$Public) <- list("FALSE" = "Researcher", "TRUE" = "Public")
recapture$Public %<>% as.logical()

is.na(recapture$TBarTag1Recap[!is.na(recapture$TBarTag1Recap) & recapture$TBarTag1Recap < 0]) <- TRUE
is.na(recapture$TBarTag2Recap[!is.na(recapture$TBarTag2Recap) & recapture$TBarTag2Recap < 0]) <- TRUE

recapture %<>% inner_join(select(capture, -Section), by = "Capture") %>%
  verify(DateTimeRecapture > DateTimeCapture)

recapture$TBarTag1 <- recapture$TBarTag1 %in% c(recapture$TBarTag1Recap, recapture$TBarTag2Recap)
recapture$TBarTag2 <- recapture$TBarTag2 %in% c(recapture$TBarTag1Recap, recapture$TBarTag2Recap)

recapture %<>% arrange(DateTimeRecapture)
recapture %<>% select(DateTimeRecapture, Capture, SectionRecapture = Section,
                      TBarTag1, TBarTag2,
                      TagsRemoved, Released, Public)

capture %<>% arrange(Species, DateTimeCapture, Capture)
capture$Capture %<>% sprintf("%03d", .) %>% paste0("F", .) %>% factor(., levels = .)
recapture$Capture %<>% sprintf("%03d", .) %>% paste0("F", .) %>% factor(., levels = levels(capture$Capture))

fillin_missing <- function(x, bol) {
  x_name <- deparse(substitute(x))
  message("replacing ", sum(is.na(x)), " missing values in ", x_name, " with ", bol)
  x[is.na(x)] <- bol
  x
}

recapture$Public %<>% fillin_missing(TRUE)
recapture$Released %<>% fillin_missing(FALSE)
recapture$TagsRemoved %<>% fillin_missing(TRUE)

# recapture most have at least one TBar tag
recapture %<>% filter(TBarTag1 | TBarTag2)

recapture %<>% arrange(DateTimeRecapture)

lexr:::plot_lex_recapture(recapture)
lexr:::check_lex_recapture(recapture)
use_data(recapture, overwrite = TRUE)

capture %<>% left_join(select(filter(recapture, !Released), DateTimeTagRemoved = DateTimeRecapture, Capture), by = c("Capture"))

bol <- is.na(capture$DateTimeTagRemoved) | capture$DateTimeTagRemoved > capture$DateTimeTagExpire
capture$DateTimeTagRemoved[bol] <-  capture$DateTimeTagExpire[bol]

detection <- read_csv(file.path(dir, "qryKLESVueDetectionRaw01Feb2016.txt"))

detection %<>% mutate(
  DateTimeDetection = ISOdate(YearUTC, MonthUTC, DayUTC, HourUTC, tz = "UTC")) %>%
  mutate(DateTimeDetection = with_tz(DateTimeDetection, tz_analysis))

detection %<>% filter(DateTimeDetection >= firstDateTime)

detection %<>% rename(AcousticTag = TagIDNbr,
                      Detections = DetectCount,
                      DetectionDBID = VR2WDetectID)

detection %<>% mutate(Receiver = RecNbr)

detection %<>% inner_join(capture, by = "AcousticTag")

detection %<>% filter(DateTimeDetection > DateTimeCapture, DateTimeDetection < DateTimeTagRemoved)

detection %<>% filter(Detections >= 3)

detection$Receiver %<>% factor(., levels = levels(deployment$Receiver))
detection %<>% filter(!is.na(Receiver))
detection %<>% inner_join(deployment, by = "Receiver")
detection %<>% filter(DateTimeDetection > DateTimeReceiverIn, DateTimeDetection < DateTimeReceiverOut)

detection %<>% select(DateTimeDetection, Capture, Receiver, Detections)

# filter duplicate detections
detection <- detection[!duplicated(detection[c("DateTimeDetection", "Capture", "Receiver", "Detections")]),]

lexr:::plot_lex_detection(detection)
lexr:::check_lex_detection(detection)
use_data(detection, overwrite = TRUE)

capture %<>% select(Capture, DateTimeCapture, Section, Species, Length, Weight,
                    Reward1 = TBarTag1Reward, Reward2 = TBarTag2Reward,
                    DateTimeTagExpire)
capture %<>% rename(SectionCapture = Section)

lexr:::plot_lex_capture(capture)
lexr:::check_lex_capture(capture)
use_data(capture, overwrite = TRUE)

section <- section[!section@data$SectionNumber %in% 34:38,]

section <- section[section@data$Section,]

section@data$Habitat <- factor("Lentic", levels = c("Lentic", "Lotic"))
section@data$Habitat[section@data$Section %in% c("S01", "S02", "S05", "S06", "S19", "S20", "S33")] <- "Lotic"

section@data$Bounded <- TRUE
section@data$Bounded[section@data$Section %in% c("S19", "S33")] <- FALSE

section@data %<>% select(Section, Habitat, Bounded, EastingSection, NorthingSection)
row.names(section) <- as.character(section@data$Section)

lexr:::plot_lex_section(section)
lexr:::check_lex_section(section)
use_data(section, overwrite = TRUE)
